### PR TITLE
Docstrings

### LINKED ISSUE #4: Docstrings/ User and Developer Documentation

### PR DESCRIPTION
**Pull request by mlaures**
written with Doxygen in mind. comments on if we want or need to build with doxygen for html pages?

### CODE REVIEW

**Comment by yupinghuang** [comment]
I do want html if possible.

File: mailbox.hpp [hunk 1]
```diff
@@ -86,7 +86,7 @@ void mailbox<T>::queue (int key, buffer_ptr<T> ptr) {
 
     } else {
         box.emplace(key, std::make_shared<map_value>(max_read, ptr));
-        box[key]->val_cv.notify_one();
+        box[key]->val_cv.notify_all();
```
Comment (yupinghuang) [inline mailbox.hpp:89]: put this in a separate commit?

Comment (mlaures) [reply mailbox.hpp:89]: Yea that should probably be in the other PR

**Comment by yupinghuang** [comment]
LGTM! I'm a little worried about the repo bloating from all the htmls.

If you can build the docs during the github-pages build and not commit the docs to the repo I'd prefer that. But someone can sort that out and purge the html directory down the road.

Please add instructions on how to build the docs with doxygen.
